lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FileSystemPartition.cs

```diff
@@ -0,0 +1,148 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using Org.Apache.REEF.IO.FileSystem;
+using Org.Apache.REEF.IO.PartitionedData.FileSystem.Parameters;
+using Org.Apache.REEF.IO.PartitionedData.Random.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IO.PartitionedData.FileSystem
+{
+    internal sealed class FileSystemPartition<T> : IPartition<IEnumerable<T>>, IDisposable
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(FileSystemPartition<T>));
+
+        private readonly string _id;
+        private readonly IFileSystem _fileSystem;
+        private readonly IFileDeSerializer<T> _fileSerializer;
+        private readonly ISet<string> _filePaths;
+        private bool _isInitialized;
+        private readonly object _lock = new object();
+        private string _localFileFolder;
+
+        [Inject]
+        private FileSystemPartition([Parameter(typeof (PartitionId))] string id,
+            [Parameter(typeof (FilePathsInPartition))] ISet<string> filePaths,
+            IFileSystem fileSystem,
+            IFileDeSerializer<T> fileSerializer)
+        {
+            _id = id;
+            _fileSystem = fileSystem;
+            _fileSerializer = fileSerializer;
+            _filePaths = filePaths;
+            _isInitialized = false;
+        }
+
+        public string Id
+        {
+            get { return _id; }
+        }
+
+        private void Initialize()
+        {
+            lock (_lock)
+            {
+                if (!_isInitialized)
+                {
+                    CopyFromRemote();
+                    _isInitialized = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method copy remote files to local and then deserialize the files.
+        /// It returns the IEnumerble of T, the details is defined in the Deserialize() method 
+        /// provided by the Serializer
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<T> GetPartitionHandle()
+        {
+            if (!_isInitialized)
+            {
+                Initialize();
+            }
+            return _fileSerializer.Deserialize(_localFileFolder);
```

Comment:
this needs a lock. Even with lock in Initialize, we could end up double initializing.